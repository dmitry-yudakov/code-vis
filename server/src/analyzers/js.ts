import {
  FileIncludeInfo,
  FunctionCallInfo,
  FileMapping,
  FunctionDeclarationInfo,
} from './../types';
import ts, { CallExpression, SyntaxKind } from 'typescript';

const extractIncludes = (relativePath: string, content: string) => {
  const includes: FileIncludeInfo[] = [];
  const re = /^(\s*)import (.+) from ['"](\..+)['"]/gm;
  // console.log('Analyze', relativePath, content);

  do {
    let out = re.exec(content);
    if (!out) break;
    const [, , what, whereFrom] = out;
    // console.log('include with import', { what, whereFrom });
    const whatSplit = what.split(/[,\s{}]+/).filter((t) => !!t);
    includes.push({
      items: whatSplit,
      to: relativePath,
      from: whereFrom,
    });
  } while (1);

  const re2 = /^(\s*)(const|let|var) (.+) = require\(['"](\..+)['"]\)/gm;

  do {
    let out = re2.exec(content);
    // console.log('include with require', out);
    if (!out) break;
    const [, , , what, whereFrom] = out;
    // console.log('include with require', { what, whereFrom });
    const whatSplit = what.split(/[,\s{}]+/).filter((t) => !!t);
    includes.push({
      items: whatSplit,
      to: relativePath,
      from: whereFrom,
    });
  } while (1);
  return includes;
};

const resolveRelativeIncludePathInPlace = (info: FileIncludeInfo): void => {
  const re = /\//;
  const { to, from } = info;
  const pathTokens = to.split(re).filter((t) => !!t);
  pathTokens.pop(); // remove filename and leave only path

  const fromTokens = from.split(re).filter((t) => !!t);

  for (const token of fromTokens) {
    if (token === '.') {
      // noop
    } else if (token === '..') {
      pathTokens.pop();
    } else {
      pathTokens.push(token);
    }
  }
  // info.from = '/' + pathTokens.join('/');
  info.from = pathTokens.join('/');
  // console.log(info.from, info.to);
};

const autoAppendJSextensionInPlace = (
  info: FileIncludeInfo,
  projectFiles: string[]
): void => {
  const { from } = info;
  for (let filename of projectFiles) {
    if (
      filename.length > from.length &&
      filename.length - from.length <= 5 &&
      filename.indexOf(from) === 0 &&
      filename[from.length] === '.'
    ) {
      console.log('AutoComplete', from, 'to', filename);
      info.from = filename;
      break;
    }
  }
};

const extractFilesHierarchy = async (
  filenames: string[],
  getFileContent: (filename: string) => Promise<string>
): Promise<FileIncludeInfo[]> => {
  const includes: FileIncludeInfo[] = await Promise.all(
    filenames.map(async (filename) => {
      const content = await getFileContent(filename);
      return extractIncludes(filename, content);
    })
  ).then((nestedIncludes) => nestedIncludes.flat());

  includes.forEach(resolveRelativeIncludePathInPlace);

  includes.forEach((info) => autoAppendJSextensionInPlace(info, filenames));

  return includes;
};

const searchFor = (
  tree: any,
  kind: SyntaxKind,
  result?: any[],
  path?: string[]
) => {
  const _result = result || [];
  const _path = path || [];
  if (_path.length > 100) {
    console.log('Too long path in searching:', _path);
    throw new Error('Too long path');
  }
  if (Array.isArray(tree)) {
    for (const val of tree) {
      searchFor(val, kind, _result, [..._path, 'ARR']);
    }
  } else if (typeof tree === 'object') {
    if (tree.kind === kind) {
      _result.push(tree);
      // return;
    }
    for (const [key, val] of Object.entries(tree)) {
      if (key === 'parent' || key === 'parseDiagnostics') continue;
      // console.log('analize obj prop', key);
      searchFor(val, kind, _result, [..._path, key]);
    }
  }
  return _result;
};

const extractFunctionDeclarations = (
  filename: string,
  sourceFile: ts.SourceFile
): FunctionDeclarationInfo[] => {
  const funcs = searchFor(sourceFile, SyntaxKind.FunctionDeclaration).map(
    (node) => {
      return {
        name: node.name.escapedText,
        filename,
        pos: node.pos,
        end: node.end,
        args: node.parameters.map((p: any) => {
          // console.log('Func argument:', p);
          return p.name.escapedText;
        }),
      };
    }
  );

  const arrowFuncs = searchFor(sourceFile, SyntaxKind.ArrowFunction)
    .filter((node) => {
      // console.log('Arrow node', node);
      // console.log('Arrow node parent', node.parent);
      if (node.parent?.name?.escapedText) return true;
      console.log('Unexpected arrow func declaration:', node);
    })
    .map((node) => {
      // try to get 'const ...;' parts in pos-end too.
      // sometimes impossible -> const a=()=>{}, b=()=>{};
      // usually it's sole declaration: const a = () => {}
      const isSoleDeclaration =
        node.parent.kind === SyntaxKind.VariableDeclaration &&
        node.parent.parent.kind === SyntaxKind.VariableDeclarationList &&
        node.parent.parent.declarations.length === 1 &&
        node.parent.parent.parent.kind === SyntaxKind.VariableStatement;
      const furthestRelevantParentNode = isSoleDeclaration
        ? node.parent.parent.parent
        : node.parent;
      // console.log({ isSoleDeclaration, furthestRelevantParentNode });
      return {
        name: node.parent.name.escapedText,
        filename,
        pos: furthestRelevantParentNode.pos,
        end: furthestRelevantParentNode.end,
        args: node.parameters.map((p: any) => {
          // console.log('Func argument:', p);
          return p.name.escapedText;
        }),
      };
    });

  const methods = searchFor(sourceFile, SyntaxKind.MethodDeclaration)
    .filter((node) => {
      // console.log('Arrow node', node);
      // console.log('Arrow node parent', node.parent);
      if (node.parent?.name?.escapedText) return true;
      console.log('Unexpected method declaration:', node);
    })
    .map((node) => {
      return {
        name: node.name.escapedText,
        filename,
        pos: node.pos,
        end: node.end,
        args: node.parameters.map((p: any) => {
          // console.log('Func argument:', p);
          return p.name.escapedText;
        }),
      };
    });

  return [...funcs, ...arrowFuncs, ...methods];
};

const extractFunctionCalls = (
  filename: string,
  sourceFile: ts.SourceFile
): FunctionCallInfo[] => {
  const calls = searchFor(sourceFile, SyntaxKind.CallExpression)
    .filter((node: any) => {
      if (!!node.arguments && !!node.expression?.escapedText) return true;
      if (!!node.arguments && !!node.expression?.name?.escapedText) return true;
      if (node.expression?.kind === SyntaxKind.SuperKeyword) return false;
      console.log('Unexpected CallExpression:', node);
    })
    .map((node: any) => {
      const name =
        node.expression.escapedText || node.expression?.name?.escapedText;
      const args = node.arguments.map((arg: any) => {
        // console.log('Arg', arg);
        // console.log('Arg expr', arg.expression);
        return arg.text || `EXPR:${arg.expression?.escapedText}...` || 'n/a';
      });

      return {
        name,
        args,
        pos: node.pos,
        end: node.end,
        filename,
      };
    });

  return calls;
};

const extractFileMapping = (
  relativePath: string,
  content: string
): FileMapping => {
  const sourceFile = ts.createSourceFile(
    relativePath,
    content,
    ts.ScriptTarget.Latest,
    true
  );
  // console.log(sourceFile);

  const includes = extractIncludes(relativePath, content);

  const functionDeclarations: FunctionDeclarationInfo[] = extractFunctionDeclarations(
    relativePath,
    sourceFile
  );

  const functionCalls: FunctionCallInfo[] = extractFunctionCalls(
    relativePath,
    sourceFile
  );

  const res = { functionDeclarations, includes, functionCalls };
  // console.log('res', JSON.stringify(res, null, 2));
  return res;
};

export default {
  extractFilesHierarchy,
  extractFileMapping,
};
