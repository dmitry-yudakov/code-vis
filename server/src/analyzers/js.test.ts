import { FileMapping, FunctionCallInfo } from '../types';
import jsAnalyzer from './js';

describe('Includes using "import"', () => {
  test('from same directory', () =>
    expect(
      jsAnalyzer.extractFilesHierarchy(['src/dir/a.js'], async (filename) => {
        expect(filename).toBe('src/dir/a.js');
        return `import gaga from './b.js'`;
      })
    ).resolves.toEqual([
      { from: 'src/dir/b.js', to: 'src/dir/a.js', items: ['gaga'] },
    ]));

  test('from different subdirectory', () =>
    expect(
      jsAnalyzer.extractFilesHierarchy(
        ['src/dir/a.js'],
        async () => `import gaga from '../dir2/b.js'`
      )
    ).resolves.toEqual([
      { from: 'src/dir2/b.js', to: 'src/dir/a.js', items: ['gaga'] },
    ]));

  test('from different directory', () =>
    expect(
      jsAnalyzer.extractFilesHierarchy(
        ['src/dir/a.js'],
        async () => `import gaga from '../../src2/dir2/b.js'`
      )
    ).resolves.toEqual([
      { from: 'src2/dir2/b.js', to: 'src/dir/a.js', items: ['gaga'] },
    ]));

  test('non-default exports', () =>
    expect(
      jsAnalyzer.extractFilesHierarchy(
        ['src/dir/a.js'],
        async () => `import { gaga, maga } from './b.js'`
      )
    ).resolves.toEqual([
      { from: 'src/dir/b.js', to: 'src/dir/a.js', items: ['gaga', 'maga'] },
    ]));

  test('mixed exports', () =>
    expect(
      jsAnalyzer.extractFilesHierarchy(
        ['src/dir/a.js'],
        async () => `import { gaga, maga }, yaga from './b.js'`
      )
    ).resolves.toEqual([
      {
        from: 'src/dir/b.js',
        to: 'src/dir/a.js',
        items: ['gaga', 'maga', 'yaga'],
      },
    ]));

  test('multiple imports', () =>
    expect(
      jsAnalyzer.extractFilesHierarchy(
        ['src/dir/a.js'],
        async () => `
      import { gaga, maga } from './b.js';
      import djaga from './c.js';
      `
      )
    ).resolves.toEqual([
      {
        from: 'src/dir/b.js',
        to: 'src/dir/a.js',
        items: ['gaga', 'maga'],
      },
      {
        from: 'src/dir/c.js',
        to: 'src/dir/a.js',
        items: ['djaga'],
      },
    ]));

  // test('multiple lines', () =>
  //   expect(
  //     jsAnalyzer.extractFilesHierarchy(
  //       ['src/dir/a.js'],
  //       async () => `
  //     import {
  //       gaga,
  //       maga
  //     } from './b.js';
  //     `
  //     )
  //   ).resolves.toEqual([
  //     {
  //       from: 'src/dir/b.js',
  //       to: 'src/dir/a.js',
  //       items: ['gaga', 'maga'],
  //     },
  //   ]));
});

describe('Includes using "require"', () => {
  test('from same directory', () =>
    expect(
      jsAnalyzer.extractFilesHierarchy(['src/dir/a.js'], async (filename) => {
        expect(filename).toBe('src/dir/a.js');
        return `const gaga = require('./b.js')`;
      })
    ).resolves.toEqual([
      { from: 'src/dir/b.js', to: 'src/dir/a.js', items: ['gaga'] },
    ]));

  test('from different subdirectory', () =>
    expect(
      jsAnalyzer.extractFilesHierarchy(
        ['src/dir/a.js'],
        async () => `const gaga = require('../dir2/b.js')`
      )
    ).resolves.toEqual([
      { from: 'src/dir2/b.js', to: 'src/dir/a.js', items: ['gaga'] },
    ]));

  test('from different directory', () =>
    expect(
      jsAnalyzer.extractFilesHierarchy(
        ['src/dir/a.js'],
        async () => `const gaga = require('../../src2/dir2/b.js')`
      )
    ).resolves.toEqual([
      { from: 'src2/dir2/b.js', to: 'src/dir/a.js', items: ['gaga'] },
    ]));

  test('non-default exports', () =>
    expect(
      jsAnalyzer.extractFilesHierarchy(
        ['src/dir/a.js'],
        async () => `const { gaga, maga } = require('./b.js')`
      )
    ).resolves.toEqual([
      { from: 'src/dir/b.js', to: 'src/dir/a.js', items: ['gaga', 'maga'] },
    ]));

  test('multiple requires', () =>
    expect(
      jsAnalyzer.extractFilesHierarchy(
        ['src/dir/a.js'],
        async () => `
      const { gaga, maga } = require('./b.js');
      const djaga = require('./c.js');
      `
      )
    ).resolves.toEqual([
      {
        from: 'src/dir/b.js',
        to: 'src/dir/a.js',
        items: ['gaga', 'maga'],
      },
      {
        from: 'src/dir/c.js',
        to: 'src/dir/a.js',
        items: ['djaga'],
      },
    ]));

  test('var/let/const', () =>
    expect(
      jsAnalyzer.extractFilesHierarchy(
        ['src/dir/a.js'],
        async () => `
        const gaga = require('./b.js');
        let djaga = require('./c.js');
        var yaga = require('./d.js');
        `
      )
    ).resolves.toEqual([
      {
        from: 'src/dir/b.js',
        to: 'src/dir/a.js',
        items: ['gaga'],
      },
      {
        from: 'src/dir/c.js',
        to: 'src/dir/a.js',
        items: ['djaga'],
      },
      {
        from: 'src/dir/d.js',
        to: 'src/dir/a.js',
        items: ['yaga'],
      },
    ]));

  // test('multiple lines', () =>
  //   expect(
  //     jsAnalyzer.extractFilesHierarchy(
  //       ['src/dir/a.js'],
  //       async () => `
  //     const {
  //       gaga,
  //       maga
  //     } = require('./b.js');
  //     `
  //     )
  //   ).resolves.toEqual([
  //     {
  //       from: 'src/dir/b.js',
  //       to: 'src/dir/a.js',
  //       items: ['gaga', 'maga'],
  //     },
  //   ]));
});

describe('File mapping', () => {
  test('regular functions', () => {
    const content = `
const somevar = 42;

function gaga(a) {
  return a + a;
}

gaga(42);
const c = 3 + gaga(12) + FEFE(gaga(33));
    `;
    const res = jsAnalyzer.extractFileMapping('src/dir/a.js', content);
    // console.log(content, JSON.stringify(res, null, 2));
    expect(res).toMatchSnapshot();
  });

  test('arrow functions', () => {
    const content = `
const somevar = 42;

const maga = (a: string, b: string) => a + b;
const yaga = (a: number, b: any) => {
  const c = a + b;
  return c * 2;
}
maga(1,2);
yaga(1,2);
    `;
    const res = jsAnalyzer.extractFileMapping('src/dir/a.js', content);
    // console.log(content, JSON.stringify(res, null, 2));
    expect(res).toMatchSnapshot();
  });

  test('class functions', () => {
    const content = `
class CL {
  constructor() {
    super();
  };
  propFunc = (a) => a;
  methodFunc(a) {
    return a;
  }
}
const cl = new CL();
cl.propFunc(42);
cl.methodFunc(12);
    `;
    const res = jsAnalyzer.extractFileMapping('src/dir/a.js', content);
    // console.log(content, JSON.stringify(res, null, 2));
    expect(res).toMatchSnapshot();
  });
});
