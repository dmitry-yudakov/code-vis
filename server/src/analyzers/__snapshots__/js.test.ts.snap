// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File mapping arrow functions 1`] = `
Object {
  "functionCalls": Array [
    Object {
      "args": Array [
        "1",
        "2",
      ],
      "end": 152,
      "filename": "src/dir/a.js",
      "name": "maga",
      "pos": 142,
    },
    Object {
      "args": Array [
        "1",
        "2",
      ],
      "end": 163,
      "filename": "src/dir/a.js",
      "name": "yaga",
      "pos": 153,
    },
  ],
  "functionDeclarations": Array [
    Object {
      "args": Array [
        "a",
        "b",
      ],
      "end": 66,
      "filename": "src/dir/a.js",
      "name": "maga",
      "pos": 34,
    },
    Object {
      "args": Array [
        "a",
        "b",
      ],
      "end": 142,
      "filename": "src/dir/a.js",
      "name": "yaga",
      "pos": 80,
    },
  ],
  "includes": Array [],
}
`;

exports[`File mapping class functions 1`] = `
Object {
  "functionCalls": Array [
    Object {
      "args": Array [
        "42",
      ],
      "end": 145,
      "filename": "src/dir/a.js",
      "name": "propFunc",
      "pos": 129,
    },
    Object {
      "args": Array [
        "12",
      ],
      "end": 164,
      "filename": "src/dir/a.js",
      "name": "methodFunc",
      "pos": 146,
    },
  ],
  "functionDeclarations": Array [
    Object {
      "args": Array [
        "a",
      ],
      "end": 69,
      "filename": "src/dir/a.js",
      "name": "propFunc",
      "pos": 60,
    },
    Object {
      "args": Array [
        "a",
      ],
      "end": 106,
      "filename": "src/dir/a.js",
      "name": "methodFunc",
      "pos": 70,
    },
  ],
  "includes": Array [],
}
`;

exports[`File mapping regular functions 1`] = `
Object {
  "functionCalls": Array [
    Object {
      "args": Array [
        "42",
      ],
      "end": 68,
      "filename": "src/dir/a.js",
      "name": "gaga",
      "pos": 58,
    },
    Object {
      "args": Array [
        "12",
      ],
      "end": 92,
      "filename": "src/dir/a.js",
      "name": "gaga",
      "pos": 83,
    },
    Object {
      "args": Array [
        "EXPR:gaga...",
      ],
      "end": 109,
      "filename": "src/dir/a.js",
      "name": "FEFE",
      "pos": 94,
    },
    Object {
      "args": Array [
        "33",
      ],
      "end": 108,
      "filename": "src/dir/a.js",
      "name": "gaga",
      "pos": 100,
    },
  ],
  "functionDeclarations": Array [
    Object {
      "args": Array [
        "a",
      ],
      "end": 58,
      "filename": "src/dir/a.js",
      "name": "gaga",
      "pos": 20,
    },
  ],
  "includes": Array [],
}
`;
